neo/Consensus/ConsensusService.cs

```diff
@@ -228,6 +251,133 @@ private void OnPersistCompleted(Block block)
             InitializeConsensus(0);
         }
 
+        private (ConsensusPayload, PrepareRequest) ReverifyPrepareRequest(ConsensusContext consensusContext, RecoveryMessage message)
+        {
+            PrepareRequest prepareRequest = new PrepareRequest
+            {
+                Nonce = message.Nonce,
+                NextConsensus = message.NextConsensus,
+                TransactionHashes = message.TransactionHashes,
+                MinerTransaction = message.MinerTransaction
+            };
+            return (consensusContext.RegenerateSignedPayload(prepareRequest, (ushort) consensusContext.PrimaryIndex,
+                message.PrepareMsgWitnessInvocationScripts[consensusContext.PrimaryIndex]), prepareRequest);
+        }
+
+        private void OnRecoveryMessageReceived(ConsensusPayload payload, RecoveryMessage message)
+        {
+            if (context.State.HasFlag(ConsensusState.CommitSent)) return;
+            if (context.BlockIndex > payload.BlockIndex) return;
+            if (context.ViewNumber > message.ViewNumber) return;
+            Snapshot snap =  Blockchain.Singleton.GetSnapshot();
+            if (payload.BlockIndex > snap.Height + 1) return;
+
+            Log($"{nameof(OnRecoveryMessageReceived)}: height={payload.BlockIndex} view={message.ViewNumber} index={payload.ValidatorIndex}");
+
+            if (context.ViewNumber == message.ViewNumber)
+            {
+                // if we are already on the right view number the only thing we might want to do is accept more
+                // Preparation messages if they can be reconstructed from the regeneration message.
+                if (message.PrepareMsgWitnessInvocationScripts[context.PrimaryIndex] == null) return;
+                var (prepareRequestPayload, prepareRequest) = ReverifyPrepareRequest((ConsensusContext) context, message);
+                if (!prepareRequestPayload.Verify(snap)) return;
+
+                for (int i = 0; i < context.Validators.Length; i++)
+                {
+                    if (context.Preparations[i] != null) continue;
+                    if (message.PrepareMsgWitnessInvocationScripts[i] == null) continue;
+                    if (i == context.PrimaryIndex)
+                    {
+                        OnPrepareRequestReceived(prepareRequestPayload, prepareRequest);
+                        continue;
+                    }
+
+                    var prepareResponseMsg = new PrepareResponse { PreparationHash = prepareRequestPayload.Hash };
+                    var regeneratedPrepareResponse = ((ConsensusContext) context).RegenerateSignedPayload(
```

Comment:
After we get the regenerated messages, can we just put them into the message pool and handle them in `OnReceive()`?

Reply:
We could, but it only saves about 2 lines of code at the most, and it is more overhead.

Reply:
There are some extra checks in `OnConsensusPayload()` that cannot be skipped.

Reply:
Maybe we should just do the checks, otherwise it re-deserializes the payload.

Reply:
Ok, I took care of it.